### PR TITLE
containerapp: Fail when explicit Dockerfile path not found

### LINKED ISSUE #4086: When the `Dockerfile` for ACA builds isn't found, the error is not obvious
- [x] Make sure you've installed the latest version using [instructions in the wiki](../wiki/install)

**Output from `azd version`**
Run `azd version` and copy and paste the output here:

`azd version 1.9.4 (commit 60d7a770c73289e303a539babf5965e638843227)`

**Describe the bug**
I did a refactor of a codebase which saw the need to move the `Dockerfile`s that services used, so I updated the `azure.yaml` file to reflect the paths, but made them relative to the location of the `azure.yaml` file rather than the path of the service.

When running `azd pack`, the service failed to build due to #3124, since the `AzureAIProxy.Shared` project didn't get copied in by Oryx, since it's a sibling of the service path.

This results in an error that is unrelated to the actual error (my supplied `Dockerfile` not being found), causing me spend time going down the wrong pathway to try and resovle the problem. Even looking at the `--debug` logs does not make it particularly obvious as to why it wasn't using my supplied `Dockerfile`. Eventually, I found the problem through inspecting the azd source code to work out why Oryx was used and not my supplied `Dockerfile`, resulting in a simple fix:

```diff
diff --git a/azure.yaml b/azure.yaml
index 7013621..1548038 100644
--- a/azure.yaml
+++ b/azure.yaml
@@ -7,8 +7,8 @@ services:
     language: py
     host: containerapp
     docker:
-      path: ./src/Dockerfile.proxy
-      context: .
+      path: ../Dockerfile.proxy
+      context: ../
       platform: amd64
 
   playground:
@@ -28,8 +28,8 @@ services:
         interactive: false
         run: ../../infra/scripts/setup-app-redirect-uris.sh
     docker:
-      path: ./src/Dockerfile.management
-      context: .
+      path: ../Dockerfile.management
+      context: ../
       platform: amd64
 
 hooks:
```

**To Reproduce**
* Clone [this commit](https://github.com/microsoft/azure-openai-service-proxy/commit/147660a3f025d3c4efc11c623097dfa06a7e4918)
* Run `azd pack admin` in the root of the repository
* Observe the error

**Expected behavior**

The error message should be more clear as to why the `Dockerfile` was not found and the reason that Oryx was used instead. At present, there is no real error message, it just prints **Building Docker image from source** and carries on as if nothing is wrong.

### PR DESCRIPTION
**Pull request by weikanglim**
This change fails the service build when an explicit Dockerfile path is specified using `docker.path`, and the Dockerfile isn't found.

When no Dockerfile is specified, or `docker.path` is set to empty, `azd` will continue to use Oryx build.

Fixes #4086, #3551

### CODE REVIEW

**Review by ellismg — APPROVED** [review]
LGTM

File: cli/azd/pkg/project/framework_service_docker.go [hunk 1]
```diff
@@ -254,13 +254,10 @@ func (p *dockerProject) Build(
...
 
 			_, err := os.Stat(path)
-			if err != nil && !errors.Is(err, os.ErrNotExist) {
-				task.SetError(fmt.Errorf("reading dockerfile: %w", err))
-				return
-			}
-
-			if errors.Is(err, os.ErrNotExist) {
-				// Build the container from source
+			if errors.Is(err, os.ErrNotExist) && serviceConfig.Docker.Path == "" {
```
Comment (jongio) [inline cli/azd/pkg/project/framework_service_docker.go:257]: Could you also write tests for this?

Comment (weikanglim) [reply cli/azd/pkg/project/framework_service_docker.go:257]: The success case here is covered by the test [here](https://github.com/Azure/azure-dev/blob/311d3c832cbe7f6392a632f8949b020dbdd125a6/cli/azd/pkg/project/framework_service_docker_test.go#L340). The failure case is an `err != nil { return err }` so I think we're OK not testing explicitly for the failure.

Comment (jongio) [reply cli/azd/pkg/project/framework_service_docker.go:257]: It looks like that tests when there is no docker file, not when it has an invalid docker file.  Is that true?

Comment (weikanglim) [reply cli/azd/pkg/project/framework_service_docker.go:257]: It indeed tests that no `docker.path` is specified. Are you asking to the failure case of "docker file not found", i.e. the `err != nil { return err }` case?

**Review by vhvb1989 — APPROVED** [review]

**Pull request merged**